public/get-pip.py

```diff
@@ -38,48 +38,48 @@
 import pkgutil
 import shutil
 import tempfile
+import argparse
+import importlib
 from base64 import b85decode
 
 
-def determine_pip_install_arguments():
-    implicit_pip = True
-    implicit_setuptools = True
-    implicit_wheel = True
+def include_setuptools(args):
+    """
+    Install setuptools only if absent and not excluded.
+    """
+    cli = not args.no_setuptools
+    env = not os.environ.get("PIP_NO_SETUPTOOLS")
+    try:
+        importlib.import_module("setuptools")
+        absent = False
+    except ImportError:
+        absent = True
+    return cli and env and absent
 
-    # Check if the user has requested us not to install setuptools
-    if "--no-setuptools" in sys.argv or os.environ.get("PIP_NO_SETUPTOOLS"):
-        args = [x for x in sys.argv[1:] if x != "--no-setuptools"]
-        implicit_setuptools = False
-    else:
-        args = sys.argv[1:]
 
-    # Check if the user has requested us not to install wheel
-    if "--no-wheel" in args or os.environ.get("PIP_NO_WHEEL"):
-        args = [x for x in args if x != "--no-wheel"]
-        implicit_wheel = False
+def include_wheel(args):
+    """
+    Install wheel only if absent and not excluded.
+    """
+    cli = not args.no_wheel
+    env = not os.environ.get("PIP_NO_WHEEL")
+    try:
+        importlib.import_module("wheel")
+        absent = False
+    except ImportError:
+        absent = True
+    return cli and env and absent
 
-    # We only want to implicitly install setuptools and wheel if they don't
-    # already exist on the target platform.
-    if implicit_setuptools:
-        try:
-            import setuptools  # noqa
-            implicit_setuptools = False
-        except ImportError:
-            pass
-    if implicit_wheel:
-        try:
-            import wheel  # noqa
-            implicit_wheel = False
-        except ImportError:
-            pass
 
-    # Add any implicit installations to the end of our args
-    if implicit_pip:
-        args += ["pip"]
-    if implicit_setuptools:
-        args += ["setuptools"]
-    if implicit_wheel:
-        args += ["wheel"]
+def determine_pip_install_arguments():
+    pre_parser = argparse.ArgumentParser()
+    pre_parser.add_argument("--no-setuptools", action="store_true")
+    pre_parser.add_argument("--no-wheel", action="store_true")
+    pre, args = pre_parser.parse_known_args()
+
+    args += ["pip"]
+    args += ["setuptools"] * include_setuptools(pre)
```

Comment:
This can be made a little less obscure by using a simple if statement instead:

```
if include_setuptools(pre):
    args.append("setuptools")
```

Reply:
I prefer the list math, but I'll follow your preferences.

Reply:
Fixed in 9a7363b62287169b96c329f66ecf043c21492530.